### PR TITLE
Bring up and optimize Whisper model on Wormhole/Blackhole, clean up demo and tests, delete old obsolete versions

### PR DESCRIPTION
**Pull request by skhorasganiTT**
### Ticket
N/A

### Problem description
- We did not have a functional + optimized implementation of the Whisper model on Wormhole or Blackhole (previously functional and unoptimized on Grayskull), which is being requested by customers
- There was a significant amount of untested and duplicate code for Whisper

### What's changed
- Deleted `models/experimental/whisper`, `models/experimental/functional_whisper/reference`, `models/experimental/functional_whisper/tt/ttnn_functional_whisper.py` and all corresponding tests
- Moved `models/experimental/functional_whisper` to `models/demos/whisper`
- Enabled program cache and async in all Whisper tests 
- Implemented kv caching for the model and demo, added to all module/perf tests
- Moved all ops running on torch/host to device (in particular, TMs in decoder attention and conv-1D ops in encoder)
- Added support for the newer `distil-whisper/distil-large-v3` version (in addition to the `openai/whisper-base` config)
- Fixed PCC logging in module tests
- Added perf measurements and logging to the demo
- Added accuracy evaluations to the demo (conditional generation dataset setting)
- Cleaned up demo README, added demo test to CI
- With all changes above, improved the WH decode perf for `distil-whisper/distil-large-v3` from 5 -> 47 t/s/u
- Also enabled functionality on Blackhole (no implementation differences from the new WH implementation), which currently gives the same perf as WH (Note: Blackhole Whisper CI tests will be added in a later PR)

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [x] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by skhorasganiTT** [comment]
Single card demo/nightly tests: https://github.com/tenstorrent/tt-metal/actions/runs/13664068030
Single card model perf tests: https://github.com/tenstorrent/tt-metal/actions/runs/13664782513
All post-commit tests: https://github.com/tenstorrent/tt-metal/actions/runs/13664236723

File: models/demos/whisper/demo/demo.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+            if print_each_iter:
+                logger.info(processor.batch_decode(torch.stack(output_ids, dim=1), skip_special_tokens=True)[0])
+            yield ttnn_transcription
+
+            if next_tokens == config.eos_token_id:
+                break
+
+        ttft = first_token_time - start_encode
+        total_generate_time = time.time() - start_encode
+        logger.info(f"Time to first token: {(ttft*1000):.3f}ms")
```
Comment (uaydonat) [inline models/demos/whisper/demo/demo.py:204]: should demo fail on slow e2d perf?

File: models/demos/whisper/demo/demo.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+
+        logits = ttnn.matmul(pooled_output, parameters.classifier.weight)
+        logits = ttnn.add(logits, parameters.classifier.bias)
+
+        logits_torch = ttnn.to_torch(logits)
+        predicted_class_ids = torch.argmax(logits_torch).item()
+        predicted_label = model.config.id2label[predicted_class_ids]
+
+        logger.info("predicted_label")
+        logger.info(predicted_label)
```
Comment (uaydonat) [inline models/demos/whisper/demo/demo.py:301]: should demo fail on wrong label in CI?

File: models/demos/whisper/demo/demo.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+        ttnn_output = model_pipeline(data, sampling_rate, stream=False)
+        logger.info(f"Model output: {ttnn_output}")
+
+        # Compute word and character error rates
+        reference = ds_input["text"].lower()
+        predicted = ttnn_output.lower()
+        total_wer += jiwer.wer(reference, predicted)
+        total_cer += jiwer.cer(reference, predicted)
+
+    logger.info(f"Average Word Error Rate: {total_wer / len(ds):.4f}")
```
Comment (uaydonat) [inline models/demos/whisper/demo/demo.py:401]: assert on expected error rates?

**Review by uaydonat — CHANGES_REQUESTED** [review]

**Comment by uaydonat** [comment]
Please add whisper to BH CI as well (can be done on separate PR)

**Review by bbradelTT — APPROVED** [review]

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+        memory_config=WHISPER_MEMORY_CONFIG,
+    )
+    hidden_states = hidden_states @ parameters.fc1.weight + parameters.fc1.bias
+    hidden_states = gelu(hidden_states)
+    hidden_states = dropout(hidden_states, p=0, training=False)
+    hidden_states = hidden_states @ parameters.fc2.weight + parameters.fc2.bias
+    hidden_states = dropout(hidden_states, p=0, training=False)
+    hidden_states = residual + hidden_states
+
+    # if hidden_states.dtype == torch.float16 and (torch.isinf(hidden_states).any() or torch.isnan(hidden_states).any()):
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Thanks, removed

File: models/demos/whisper/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# Whisper Demo (Wormhole)
+
+## How to Run
+
+Whisper uses 8x8 core grid size, so the following environment variable needs to be set on a N300 setup:
+
+```sh
+export WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply models/demos/whisper/README.md:8]: Added blackhole to the readme

File: CODEOWNERS [hunk 6]
```diff
@@ -177,6 +177,7 @@ models/demos/t3000/llama3_70b @cglagovichTT @uaydonat @johanna-rock-tt @djordje-
 models/demos/t3000/mixtral8x7b @yieldthought @mtairum @uaydonat
 models/demos/tg/llama3_70b @cglagovichTT @uaydonat @johanna-rock-tt @djordje-tt @kpaigwar
 models/demos/tg/falcon7b @skhorasganiTT @djordje-tt @uaydonat
+models/demos/whisper @skhorasganiTT @uaydonat
```
Comment (tt-rkim) [inline CODEOWNERS:180]: Maybe one more owner?

File: models/demos/whisper/demo/dataset/audio_classification/10116516891483200485.wav
Comment (tt-rkim) [inline models/demos/whisper/demo/dataset/audio_classification/10116516891483200485.wav:1]: Do we need `libsndfile` ?

File: models/demos/whisper/demo/demo.py (hunk 3 above)
Comment (skhorasganiTT) [reply models/demos/whisper/demo/demo.py:401]: I didn't add the dataset evaluation to demo CI since it takes a bit longer to run and should be redundant given that the conditional generation demo and pcc checks are tested

File: tt_metal/python_env/requirements-dev.txt [hunk 7]
```diff
@@ -68,3 +68,4 @@ blobfile==2.1.1 # Required for llama3
 numpy>=1.24.4,<2
 huggingface-hub==0.25.2
 pydantic==2.9.2 # Required for Superset benchmarking
+jiwer==3.0.5 # Required for speech recognition evaluation
```
Comment (tt-rkim) [inline tt_metal/python_env/requirements-dev.txt:71]: Looks like wheel download is pretty small https://pypi.org/project/jiwer/3.1.0/#files

**Review by tt-rkim — COMMENTED** [review]
I'd recommend a custom test dispatch with something like a `--count 50` to ensure that it's stable on CI VMs: https://github.com/tenstorrent/tt-metal/actions/workflows/test-dispatch.yaml

File: models/demos/whisper/demo/dataset/audio_classification/10116516891483200485.wav
Comment (skhorasganiTT) [reply models/demos/whisper/demo/dataset/audio_classification/10116516891483200485.wav:1]: Using `from scipy.io import wavfile` for reading the wav files

File: CODEOWNERS (hunk 6 above)
Comment (skhorasganiTT) [reply CODEOWNERS:180]: No one else has looked at the model so I'm not sure who else we could add

File: models/demos/whisper/demo/dataset/audio_classification/10116516891483200485.wav
Comment (skhorasganiTT) [reply models/demos/whisper/demo/dataset/audio_classification/10116516891483200485.wav:1]: Yeah think so, though I'm not seeing that in the current requirements

File: models/demos/whisper/demo/demo.py (hunk 1 above)
Comment (skhorasganiTT) [reply models/demos/whisper/demo/demo.py:211]: Added e2e perf checks to the demo

File: models/demos/whisper/demo/demo.py (hunk 2 above)
Comment (skhorasganiTT) [reply models/demos/whisper/demo/demo.py:321]: Same comment as the error rates, I think the pcc checks should be sufficient for us

**Review by uaydonat — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]
